txscript/script.go

```diff
@@ -345,10 +345,188 @@ func removeOpcodeByData(pkscript []parsedOpcode, data []byte) []parsedOpcode {
+	var bVersion [4]byte
+	binary.LittleEndian.PutUint32(bVersion[:], uint32(txCopy.Version))
+	sigHash.Write(bVersion[:])
+
+	// Next write out the the possibly pre-calculated hashes for the
+	// sequence numbers of all inputs, and the hashes of the previous
+	// outs for all outputs.
+	var zeroHash wire.ShaHash
+
+	// If anyone can pay isn't active, then we can use the cached
+	// hashPrevOuts, otherwise we just write zeroes for the prev outs.
+	if hashType&SigHashAnyOneCanPay == 0 {
+		sigHash.Write(sigHashes.HashPrevOuts[:])
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// If the sig hash isn't anyone can pay, single, or none, the use the
+	// cached hash sequences, otherwise write all zeroes for the
+	// hashSequence.
+	if hashType&SigHashAnyOneCanPay == 0 &&
+		hashType&sigHashMask != SigHashSingle &&
+		hashType&sigHashMask != SigHashNone {
+		sigHash.Write(sigHashes.HashSequence[:])
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// Next, write the oint being spent.
+	sigHash.Write(txCopy.TxIn[idx].PreviousOutPoint.Hash[:])
+	var bIndex [4]byte
+	binary.LittleEndian.PutUint32(bIndex[:], txCopy.TxIn[idx].PreviousOutPoint.Index)
+	sigHash.Write(bIndex[:])
+
+	// Remove all instances of OP_CODESEPARATOR from the script.
+	// TODO(roasbeef): have removeOpcode return index of last found opcode,
+	// or just create new func?
+	script := removeOpcode(subScript, OP_CODESEPARATOR)
+	rawScript, _ := unparseScript(script)
+
+	// TODO(roasbeef): check before removing code seps?
+	if isWitnessPubKeyHash(script) {
+		// The script code for a p2wkh is a length prefix varint for
+		// the next 25 bytes, followed by a re-creation of the original
+		// p2pkh pk script.
+		sigHash.Write([]byte{0x19})
+		sigHash.Write([]byte{OP_DUP})
+		sigHash.Write([]byte{OP_HASH160})
+		sigHash.Write([]byte{OP_DATA_20})
+		sigHash.Write(rawScript[2:22])
+		sigHash.Write([]byte{OP_EQUALVERIFY})
+		sigHash.Write([]byte{OP_CHECKSIG})
+	} else {
+		// For p2wsh outputs, and future outputs, the script code is the
+		// original script, with all code separators removed, serialized
+		// with a var int length prefix.
+		wire.WriteVarBytes(&sigHash, 0, rawScript)
+	}
+
+	// Next, add the input amount, and sequence number of the input being
+	// signed.
+	var bAmount [8]byte
+	binary.LittleEndian.PutUint64(bAmount[:], uint64(amt))
+	sigHash.Write(bAmount[:])
+	var bSequence [4]byte
+	binary.LittleEndian.PutUint32(bSequence[:], txCopy.TxIn[idx].Sequence)
+	sigHash.Write(bSequence[:])
+
+	// If the current signature mode isn't single, or none, then we can
+	// re-use the pre-generated hashoutputs sighash fragment. Otherwise,
+	// we'll serialize and add only the target output index to the signature
+	// pre-image.
+	if hashType&SigHashSingle != SigHashSingle &&
+		hashType&SigHashNone != SigHashNone {
+		sigHash.Write(sigHashes.HashOutputs[:])
+	} else if hashType&sigHashMask == SigHashSingle { // TODO(roasbeef): verify range above
```

Comment:
This doesn't have the same behavior as the code in the BIP.

In particular, the code in the BIP has:
```C++
 } else if ((nHashType & 0x1f) == SIGHASH_SINGLE && nIn < txTo.vout.size()) {
      CHashWriter ss(SER_GETHASH, 0);
      ss << txTo.vout[nIn];
      hashOutputs = ss.GetHash();
  }
```

The code here does match for two reasons:

1) The BIP code ends up adding the **hash** of the serialized txout to the final calculated signature hash whereas this code is writing the entire serialized txout.
2) The BIP code will result in `hashOutputs` being the 32-byte zero hash when idx is out of range whereas this code will write nothing in that case.